# Exploring eBay Car Sales Data

In this dataset, we will be working with a dataset of used cars from eBay Kleinanzeigan, a classifieds section of the German eBay website. There are 50,000 data points sampled so that the code can run quickly.  This dataset was dirtied a bit to more closely resemble what you would expect from a scraped dataset/

In [3]:
# The below is to find the relative path where Jupyter can open up the dataset.  
# This will help you find where to place your dataset you want to read.
import os
cwd = os.getcwd()
files = os.listdir(cwd)
print("Files in %r: %s" % (cwd, files))


Files in 'C:\\Users\\kdawg\\Desktop\\python_work\\Dataquest': ['.ipynb_checkpoints', 'AppleStore.csv', 'autos.csv', 'Course 1 - Python for Data Science Fundamentals Part I + II', 'Course 2 - Python for Data Science Intermediate', 'Course 3 - Pandas and NumPy Fundamentals', 'Course 4 - Data Visualization Fundamentals', 'Course 5 - Storytelling Data Visualization and Information Design', 'Course 6 - Data Cleaning and Analysis', 'euro-daily-hist_1999_2020.csv', 'Exploring eBay Car Sales Data.ipynb', 'Exploring Hacker News Posts.ipynb', 'Finding Heavy Traffic Indicators on I-94.ipynb', 'Finished Guided Projects', 'googleplaystore.csv', 'Guided Project 1.ipynb', 'Guided Project 2 Hacker News', 'hacker_news.csv', 'Keep Working On', 'Metro_Interstate_Traffic_Volume.csv', 'Notebook', 'Storytelling Data Visualization on Exchange Rates-checkpoint.ipynb', 'Storytelling Data Visualization on Exchange Rates.ipynb', 'untitled', 'Untitled.ipynb', 'Untitled1.ipynb', 'Untitled2.ipynb', 'Untitled3.ipynb

## Introduction to the dataset

In [2]:
# Importing NumPy and pandas.  Reading the dataset into Jupyter.  The dataset we have read into Jupyter is called 'autos'.

import numpy as np
import pandas as pd

autos = pd.read_csv("autos.csv", encoding="windows-1252")

In [3]:
# Since we imported pandas, jupyter is able to render the first few and last few values of the pandas object.
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

- Printing DataFrame.info() shows that there are 20 columns and 371,528 rows.  There is a mixture of 'int' and 'object' data types.
- Printing DataFrame.head() renders the first 5 rows of the dataframe.  Since this dataset is from a German website, most of the words to describe the rows are in German.  You can also see that there are some null values in columns such as 'vehicleType' and 'model'.

## Cleaning the Column Names

In [5]:
# Creating a copy of the autos dataframe.
autos_copy = autos.copy()
autos_copy.columns

# Defining a function to change most of the column names.
def rename_columns(col):
    col = col.replace("yearOfRegistration","registration_year")
    col = col.replace("monthOfRegistration","registration_month")
    col = col.replace("notRepairedDamage","unrepaired_damage")
    col = col.replace("dateCreated","ad_created")
    col = col.replace("dateCrawled","date_crawled")
    col = col.replace("offerType","offer_type")
    col = col.replace("abtest","ab_test")
    col = col.replace("vehicleType","vehicle_type")
    col = col.replace("powerPS","power_PS")
    col = col.replace("fuelType","fuel_type")
    col = col.replace("nrOfPictures","nr_of_pictures")
    col = col.replace("postalCode","postal_code")
    col = col.replace("lastSeen","last_seen")
    return col

# Copying the 'autos.copy' columns names to the real 'autos' dataframe. 
renamed_columns = []
for c in autos.columns:
    renamed_col = rename_columns(c)
    renamed_columns.append(renamed_col)
autos.columns = renamed_columns
autos.head()


date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_PS   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In the cell above, I have created a copy 'autos' to change the names of the columns.  These name changes will make the column labels more descriptive.  We have also changed the camelcased columns labels into snakecase.  Instead of each word starting with an upper case letter, all the labels are lowercase with underscores between each word.  After making the changes, I had to replace the column labels from 'autos' with the column labels from 'autos_copy'.

In [6]:
# The describe method shows descriptive statistics for the columns that are 'int' dtypes.
autos.describe()
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_PS   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [7]:
print(autos['seller'].unique())
print(autos['nr_of_pictures'].unique())
print(autos['offer_type'].unique())

['privat' 'gewerblich']
[0]
['Angebot' 'Gesuch']


Above:  
- Three columns are candidates to get dropped for basically having a single value: 'seller', 'nr_of_pictures', and 'offer_type'.  'seller' only has 'privat' and 'nan' as values, 'nr_of_pictures' only has '0' and 'nan' as values, and 'offer_type' only has 'Angebot' and 'nan' as values.
- Two columns with numeric data stored as text are the 'price' and 'odometer' columns.  Those two columns need to have their non-numeric characters removed and need to be converted into numeric dtypes.  

## Dropping columns with mostly one value

In [8]:
autos = autos.drop(['seller','nr_of_pictures','offer_type'], axis=1)

In [9]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

    price  ab_test vehicle_type  registration_year    gearbox  power_PS  \
0  $5,000  control          bus               2004    manuell       158   
1  $8,500  control    limousine               1997  automatik       286   
2  $8,990     test    limousine               2009    manuell       102   
3  $4,350  control   kleinwagen               2007  automatik        71   
4  $1,350     test        kombi               2003    manuell         0   

    model   odometer  registration_month fuel_type       brand  \
0  andere  150,000km                   3       lpg     peugeot   
1     7er  150,000km                   6    benzin         bmw   
2    golf   70,000km                   7    benzin  volkswagen   
3  fortwo   70,000km                   6    benzin       smart   
4   focus  150,000km                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

We have now dropped the 'seller', 'nr_of_pictures', and 'offer_type' columns.  These were not helpful columns and will only clutter the dataset.

## 'Odometer' + 'Price' Converted into Numeric dtype

In [10]:
# Renaming the 'odometer' column to 'odometer_km' to make it more descriptive.
autos.rename({"odometer":"odometer_km"},axis=1,inplace=True)

# Getting rid of the comma from both the 'price' and 'odometer_km' columns, and getting rid of the $ from the prices.  Also, switching the dtype of these two columns to float.
autos["price"] = autos["price"].str.replace("$","").str.replace(",","").astype(float)
autos["odometer_km"] = autos["odometer_km"].str.replace(",","").str.replace("km","").astype(float)

<ipython-input-10-8d28a66463cd>:5: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will*not* be treated as literal strings when regex=True.
  autos["price"] = autos["price"].str.replace("$","").str.replace(",","").astype(float)


In [11]:
# Printing 3 rows from each column to show how the numeric examples work.
print(autos["price"][:3])
print(autos["odometer_km"][:3])

0    5000.0
1    8500.0
2    8990.0
Name: price, dtype: float64
0    150000.0
1    150000.0
2     70000.0
Name: odometer_km, dtype: float64


## Exploring the 'Odometer_km' and 'Price' Columns

Below are some methods to look at the 'price' and 'odometer_km' columns closer.  These methods are to make it easier to see the range of the numbers and to see what outliers there might be.  This will help us clean the data further by taking out the unrealistic or incorrect numbers.  

In [12]:
# The first line formats scientific notation. This makes it much easier to read the numbers.
pd.set_option('display.float_format', lambda x: '%.2f' % x)
print("Number of unique values in 'price':",autos["price"].unique().shape)
print(autos["price"].describe())
print("Count of specific prices for vehicles listed:\n",autos["price"].value_counts().sort_index())
autos["price"].max()

Number of unique values in 'price': (2357,)
count      50000.00
mean        9840.04
std       481104.38
min            0.00
25%         1100.00
50%         2950.00
75%         7200.00
max     99999999.00
Name: price, dtype: float64
Count of specific prices for vehicles listed:
 0.00           1421
1.00            156
2.00              3
3.00              1
5.00              2
               ... 
10000000.00       1
11111111.00       2
12345678.00       3
27322222.00       1
99999999.00       1
Name: price, Length: 2357, dtype: int64


99999999.0

In [13]:
pd.set_option('display.float_format', lambda x: '%.2f' % x)
print("All unique 'odometer_km' values:",autos["odometer_km"].unique())
print(autos["odometer_km"].describe())
print("Count of specific distances cars have driven (km):\n",autos["odometer_km"].value_counts())

All unique 'odometer_km' values: [150000.  70000.  50000.  80000.  10000.  30000. 125000.  90000.  20000.
  60000.   5000. 100000.  40000.]
count    50000.00
mean    125732.70
std      40042.21
min       5000.00
25%     125000.00
50%     150000.00
75%     150000.00
max     150000.00
Name: odometer_km, dtype: float64
Count of specific distances cars have driven (km):
 150000.00    32424
125000.00     5170
100000.00     2169
90000.00      1757
80000.00      1436
70000.00      1230
60000.00      1164
50000.00      1027
5000.00        967
40000.00       819
30000.00       789
20000.00       784
10000.00       264
Name: odometer_km, dtype: int64


Above, in the 'odometer_km' column, the [.unique()] method shows there are 13 unique values between 5000 and 150000km.  Most of the unique values are 150,000km, over 60% of the vehicles listed.

The [.unique()] method shows that the 'price' column has 2,357 unique values.  60% of the values have a price listed as $0, these will have to be removed.  We can also see that the max price listed for a vehicle is 99,999,999, this an outlier that should be taken out.  We will remove outliers from the 'price' column by creating a range of prices.  Removing the outliers from the price range should help decrease the outliers from the 'odometer_km' column as well.

## Cleaning the 'Price' and 'Odometer_km' Columns

In this section, I will be modifying the dataset to remove a range of items from the 'price' column. 

In [14]:
# Creating a copy of 'autos' because the dataset will be manipulated directly, 
# if I do something by accident I did not want, I can scrap the copy.
autos_copy = autos
autos_copy

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

         price  ab_test vehicle_type  registration_year    gearbox  power_PS  \
0      5000.00  control          bus               2004    manuell       158   
1      8500.00  control    limousine               1997  automatik       286   
2      8990.00     test    limousine               2009    manuell       102   
3      4350.00  control   kleinwagen               2007  automatik        71   
4      1350.00     test        kombi               2003    manuell         0   
...        ...      ...          ...                ...        ...       ...   
49995 24900.00  control    limousine               2011  automatik       239   
49996  1980.00  control       cabrio               1996    manuell        75   
49997 13200.00     test       cabrio               2014  automatik        69   
49998 22900.00  control        kombi               2013    manuell       150   
49999  1250.00  control    limousine               1996    manuell       101   

        model  odometer_km  registration_month fuel_type       brand  \
0      andere    150000.00                   3       lpg     peugeot   
1         7er    150000.00                   6    benzin         bmw   
2        golf     70000.00                   7    benzin  volkswagen   
3      fortwo     70000.00                   6    benzin       smart   
4       focus    150000.00                   7    benzin        ford   
...       ...          ...                 ...       ...         ...   
49995      q5    100000.00                   1    diesel        audi   
49996   astra    150000.00                   5    benzin        opel   
49997     500      5000.00                  11    benzin        fiat   
49998      a3     40000.00                  11    diesel        audi   
49999  vectra    150000.00                   1    benzin        opel   

      unrepaired_damage           ad_created  postal_code            last_seen  
0                  nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1                  nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2                  nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3                  nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4                  nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50  
...                 ...                  ...          ...                  ...  
49995              nein  2016-03-27 00:00:00        82131  2016-04-01 13:47:40  
49996              nein  2016-03-28 00:00:00        44807  2016-04-02 14:18:02  
49997              nein  2016-04-02 00:00:00        73430  2016-04-04 11:47:27  
49998              nein  2016-03-08 00:00:00        35683  2016-04-05 16:45:07  
49999              nein  2016-03-13 00:00:00        45897  2016-04-06 21:18:48  

[50000 rows x 17 columns]

In [15]:
autos_copy[autos_copy['price'].between(0,1000)] = np.nan
autos_copy[autos_copy['price'].between(100000,99999999)] = np.nan
autos_copy['price'].value_counts().sort_index()

1039.00      1
1040.00      1
1049.00      6
1050.00     95
1059.00      1
            ..
93911.00     1
94999.00     1
98500.00     1
99000.00     2
99900.00     2
Name: price, Length: 2051, dtype: int64

In [16]:
autos_copy['price'].shape
autos_copy['price'].describe()

count   37948.00
mean     7191.58
std      7920.02
min      1039.00
25%      2300.00
50%      4499.00
75%      8999.00
max     99900.00
Name: price, dtype: float64

In [17]:
autos_copy['odometer_km'].describe()

count    37948.00
mean    122576.68
std      40811.06
min       5000.00
25%     100000.00
50%     150000.00
75%     150000.00
max     150000.00
Name: odometer_km, dtype: float64

In [18]:
autos_copy['odometer_km'].value_counts(dropna=True)

150000.00    22771
125000.00     4297
100000.00     1850
90000.00      1560
80000.00      1325
70000.00      1153
60000.00      1091
50000.00       979
40000.00       794
30000.00       741
20000.00       680
5000.00        483
10000.00       224
Name: odometer_km, dtype: int64

In [19]:
autos_copy = autos
autos.describe()

price  registration_year  power_PS  odometer_km  registration_month  \
count 37948.00           37948.00  37948.00     37948.00            37948.00   
mean   7191.58            2005.77    129.32    122576.68                6.02   
std    7920.02              87.45    216.31     40811.06                3.58   
min    1039.00            1000.00      0.00      5000.00                0.00   
25%    2300.00            2001.00     80.00    100000.00                3.00   
50%    4499.00            2005.00    116.00    150000.00                6.00   
75%    8999.00            2009.00    160.00    150000.00                9.00   
max   99900.00            9999.00  17700.00    150000.00               12.00   

       postal_code  
count     37948.00  
mean      51737.20  
std       25679.40  
min        1067.00  
25%       31241.00  
50%       50933.00  
75%       72555.00  
max       99998.00

I have removed a range of prices from the 'price' column: 0 to 1,000 and 100,000 to 99,999,999.  This has removed 12,052 rows, leaving us with 37,948.  The standard deviation decreased from 481,104 to 7,920 because the incredibly large and unrealistically low prices have been removed.

I have removed that range of prices because vehicles sold for 0 to 1,000 are usually not worth looking at for serious car buyers, they can be much more trouble than they are worth.  Prices of 100,000 to 99,999,999 are unrealistically high prices for car buyers on eBay.  People with that spending power do not go to eBay to buy an expensive car.

## Exploring the date columns

There are 5 columns that should represent date values.  Some of these columns were created by the crawler, some came from the website itself.  Right now, the 'date_crawled', 'last_seen', and 'ad_created' columns are all identified as string values by pandas.

In [20]:
autos[['date_crawled','last_seen','ad_created']].head()

date_crawled            last_seen           ad_created
0  2016-03-26 17:47:46  2016-04-06 06:45:54  2016-03-26 00:00:00
1  2016-04-04 13:38:56  2016-04-06 14:45:08  2016-04-04 00:00:00
2  2016-03-26 18:57:24  2016-04-06 20:15:37  2016-03-26 00:00:00
3  2016-03-12 16:58:10  2016-03-15 03:16:28  2016-03-12 00:00:00
4  2016-04-01 14:38:50  2016-04-01 14:38:50  2016-04-01 00:00:00

The first 10 characters represent the date: YYYY-MM-DD

In [21]:
pd.set_option('display.float_format', lambda x: '%.4f' % x)
print("\nPercentages by date, sorted oldest to newest")
print(autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=True).sort_index(ascending=True))


Percentages by date, sorted oldest to newest
2016-03-05   0.0257
2016-03-06   0.0140
2016-03-07   0.0355
2016-03-08   0.0323
2016-03-09   0.0325
2016-03-10   0.0333
2016-03-11   0.0330
2016-03-12   0.0377
2016-03-13   0.0161
2016-03-14   0.0368
2016-03-15   0.0335
2016-03-16   0.0290
2016-03-17   0.0301
2016-03-18   0.0129
2016-03-19   0.0352
2016-03-20   0.0383
2016-03-21   0.0373
2016-03-22   0.0325
2016-03-23   0.0321
2016-03-24   0.0288
2016-03-25   0.0303
2016-03-26   0.0331
2016-03-27   0.0314
2016-03-28   0.0353
2016-03-29   0.0341
2016-03-30   0.0329
2016-03-31   0.0314
2016-04-01   0.0345
2016-04-02   0.0364
2016-04-03   0.0392
2016-04-04   0.0368
2016-04-05   0.0132
2016-04-06   0.0032
2016-04-07   0.0015
Name: date_crawled, dtype: float64


'date_crawled' = When this ad was first crawled.  
The website is crawled everyday, the distribution percentages do not vary too much.  The period covered is about one month.

In [22]:
pd.set_option('display.float_format', lambda x: '%.8f' % x)
print("\nPercentages by date")
print(autos['ad_created'].str[:10].value_counts(normalize=True, dropna=True).sort_index(ascending=True))


Percentages by date
2015-06-11   0.00002635
2015-08-10   0.00002635
2015-09-09   0.00002635
2015-11-10   0.00002635
2015-12-30   0.00002635
                ...    
2016-04-03   0.03952777
2016-04-04   0.03720881
2016-04-05   0.01183198
2016-04-06   0.00334668
2016-04-07   0.00131759
Name: ad_created, Length: 74, dtype: float64


'ad_created' = The date on which the eBay listing was created.                                                                 
This data covers a period of close to one year.  The number of ads created per day as time has gone by in that year has quite a bit overall, by many many times.

In [23]:
print("\nPercentages by date, sorted oldest to newest")
print(autos['last_seen'].str[:10].value_counts(normalize=True, dropna=True).sort_index(ascending=True))


Percentages by date, sorted oldest to newest
2016-03-05   0.00110678
2016-03-06   0.00355750
2016-03-07   0.00450617
2016-03-08   0.00627174
2016-03-09   0.00890693
2016-03-10   0.00969748
2016-03-11   0.01170022
2016-03-12   0.02224096
2016-03-13   0.00832718
2016-03-14   0.01206915
2016-03-15   0.01491515
2016-03-16   0.01546854
2016-03-17   0.02622009
2016-03-18   0.00735217
2016-03-19   0.01457257
2016-03-20   0.01960578
2016-03-21   0.01952672
2016-03-22   0.02079161
2016-03-23   0.01794561
2016-03-24   0.01849900
2016-03-25   0.01773479
2016-03-26   0.01589017
2016-03-27   0.01401918
2016-03-28   0.01939496
2016-03-29   0.02058079
2016-03-30   0.02350585
2016-03-31   0.02282070
2016-04-01   0.02326868
2016-04-02   0.02492885
2016-04-03   0.02455992
2016-04-04   0.02324233
2016-04-05   0.13139032
2016-04-06   0.23479498
2016-04-07   0.14058712
Name: last_seen, dtype: float64


'last_seen' = When the crawler saw this ad last online.
This data covers about one month.  The percentages are growing through the month.  This could mean that the website's use is growing, more cars are being sold over this time.  

# Exploring the 'registration_year' column

In [24]:
autos['registration_year'].describe()

count   37948.00000000
mean     2005.76918415
std        87.44775133
min      1000.00000000
25%      2001.00000000
50%      2005.00000000
75%      2009.00000000
max      9999.00000000
Name: registration_year, dtype: float64

The 'registration_year' column is whacky because the standard deviation is 87 years, the max year is the year 9,999, and the min year is the year 1,000.  This column will need to be cleaned up if we want it to help us to analyse this data.

## Dealing with Incorrect 'registration_year' Data

A car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely

In [25]:
autos_copy2 = autos

In [26]:
autos_copy2[autos_copy2['registration_year'].between(0,1900)] = np.nan
autos_copy2[autos_copy2['registration_year'].between(2017,9999)] = np.nan

In [27]:
print("Percentages by year, sorted highest to lowest:\n",autos_copy2['registration_year']
      .value_counts(normalize=True)
      .sort_index().tail(30))
print("\n")
print("Statistical details:\n",autos_copy2['registration_year'].describe())

Percentages by year, sorted highest to lowest:
 1987.00000000   0.00139478
1988.00000000   0.00284425
1989.00000000   0.00325448
1990.00000000   0.00495009
1991.00000000   0.00549706
1992.00000000   0.00563380
1993.00000000   0.00568850
1994.00000000   0.00700123
1995.00000000   0.01140435
1996.00000000   0.01370163
1997.00000000   0.02023793
1998.00000000   0.03311910
1999.00000000   0.04504307
2000.00000000   0.05231779
2001.00000000   0.05477916
2002.00000000   0.05762341
2003.00000000   0.06692192
2004.00000000   0.07069602
2005.00000000   0.07575550
2006.00000000   0.07236428
2007.00000000   0.06169835
2008.00000000   0.06022152
2009.00000000   0.05677561
2010.00000000   0.04318337
2011.00000000   0.04419527
2012.00000000   0.03563517
2013.00000000   0.02163271
2014.00000000   0.01769452
2015.00000000   0.00976343
2016.00000000   0.01687406
Name: registration_year, dtype: float64


Statistical details:
 count   36565.00000000
mean     2004.04348421
std         7.00457924
min      

In [28]:
autos = autos_copy2

When getting rid of the years 0 to 2016 and 2017 to 9999, it has changed the dataset significantly.  There are 1,383 items that fall outside of our designated dates.  Our earliest date is now 1927 and we have designated our latest date as 2016.  Most of the cars are registered between 2000 and 2010.

# Exploring Price by Brand

In [29]:
print("Statistical data of the 'brand' column:\n",autos['brand'].describe())
print("\n")
print("Number of cars per brand:\n",autos['brand'].value_counts())

Statistical data of the 'brand' column:
 count          36565
unique            40
top       volkswagen
freq            7709
Name: brand, dtype: object


Number of cars per brand:
 volkswagen        7709
bmw               4620
mercedes_benz     4112
audi              3598
opel              3235
ford              2127
renault           1342
peugeot           1008
fiat               767
skoda              701
seat               633
smart              608
toyota             538
mazda              520
citroen            505
nissan             499
mini               405
hyundai            391
sonstige_autos     375
volvo              326
kia                285
honda              268
porsche            252
mitsubishi         250
chevrolet          245
alfa_romeo         228
suzuki             211
dacia              121
chrysler           112
jeep               103
land_rover          98
jaguar              69
subaru              62
daihatsu            61
saab                50
daewoo        

There are 40 different car brands in this dataset.  Volkswagen has the highest representation by far.  

## Aggregating the brands

We will be aggregating the brands that have at least 5% of the total value in the dataset.

In [30]:
brand_counts = autos["brand"].value_counts(normalize=True)
common_brands = brand_counts[brand_counts > .05].index
print(common_brands)

Index(['volkswagen', 'bmw', 'mercedes_benz', 'audi', 'opel', 'ford'], dtype='object')


There are 6 brands that have at least 5% of the total value in the dataset.

In [31]:
aggregate_brands = {}

for b in common_brands:
    mean_price = autos.loc[autos['brand'] == b, 'price'].mean()
    aggregate_brands[b] = mean_price

In [32]:
aggregate_brands

{'volkswagen': 6743.257880399533,
 'bmw': 9045.878571428571,
 'mercedes_benz': 9269.809338521402,
 'audi': 10359.41078376876,
 'opel': 4298.587326120556,
 'ford': 5386.905500705218}

Volkswagen is the most represented car brand in the dataset. By mean price it is only more expensive than Opel and Ford cars.  The least expensive vehicle are Opel.  Audis are the third most represented car brand and the most expensive car by mean price.  

## Storing Aggregate Data in a Dataframe

For the top 6 brands, let's use aggregation to understand the average mileage for those cars and if there's any visible link with mean price.  First we need to calculate the mean mileage for each of the top brands, storing the results in a dictionary:

In [33]:
mean_mileage = {}

for b in common_brands:
    mean_price = autos.loc[autos['brand'] == b, 'odometer_km'].mean()
    mean_mileage[b] = mean_price
mean_mileage

{'volkswagen': 125475.41834219743,
 'bmw': 131909.0909090909,
 'mercedes_benz': 129992.70428015564,
 'audi': 127487.49305169539,
 'opel': 123588.87171561051,
 'ford': 119099.67089797837}

Now we have to convert both dictionaries (brands and mileage) to series objects, using the series constructor:

In [34]:
# Converting 'aggregate_brands' into a series object.
brand_series = pd.Series(aggregate_brands)
brand_series = brand_series.sort_values(ascending=False)
brand_series

audi            10359.41078377
mercedes_benz    9269.80933852
bmw              9045.87857143
volkswagen       6743.25788040
ford             5386.90550071
opel             4298.58732612
dtype: float64

In [35]:
# Converting 'mean_mileage' in a series object.
mileage_series = pd.Series(mean_mileage)
mileage_series = mileage_series.sort_values(ascending=False, inplace=False)
mileage_series

bmw             131909.09090909
mercedes_benz   129992.70428016
audi            127487.49305170
volkswagen      125475.41834220
opel            123588.87171561
ford            119099.67089798
dtype: float64

Now we have to create a datafram from the first series object using the dataframe constructor:

In [36]:
brand_dataframe = pd.DataFrame(brand_series,columns=['mean_price'])
brand_dataframe = brand_dataframe
brand_dataframe

mean_price
audi          10359.41078377
mercedes_benz  9269.80933852
bmw            9045.87857143
volkswagen     6743.25788040
ford           5386.90550071
opel           4298.58732612

Now we have to assign the other series as a new column in the dataframe:

In [37]:
# Adding a new column to 'brand_dataframe' filled with only zeroes for now.
brand_dataframe['mean_mileage_km'] = 0
brand_dataframe

mean_price  mean_mileage_km
audi          10359.41078377                0
mercedes_benz  9269.80933852                0
bmw            9045.87857143                0
volkswagen     6743.25788040                0
ford           5386.90550071                0
opel           4298.58732612                0

In [38]:
# Filling the rows of the column 'mean_mileage_km' with the data from 'mileage_series'.
brand_dataframe['mean_mileage_km'] = mileage_series


In [44]:
brand_dataframe

mean_price  mean_mileage_km
audi          10359.41078377  127487.49305170
mercedes_benz  9269.80933852  129992.70428016
bmw            9045.87857143  131909.09090909
volkswagen     6743.25788040  125475.41834220
ford           5386.90550071  119099.67089798
opel           4298.58732612  123588.87171561

The trend in this dataset shows that the cars with the higher mean prices tend to have a higher average mileage.  The top three most expensive cars: Audi, Mercedes Benz, and BMW have mileages of 127,487, 129,992, and 131,909, respectively.  The bottom three cars by mean price are Volkswagen, Ford, and Opel with mileages of 125,475, 119,099, and 123,588, respectively.  Overall the less expensive the car is, the less miles the car has.  